Services/Component/classes/class.ilComponentDefinitionReader.php

```diff
@@ -0,0 +1,122 @@
+<?php declare(strict_types=1);
+
+/* Copyright (c) 2021 ILIAS open source, Extended GPL, see docs/LICENSE */
+
+/**
+ *
+ */
+class ilComponentDefinitionReader
+{
+    /**
+     * @var ilComponentDefinitionProcessor[]
+     */
+    protected array $processors;
+
+    public function __construct(
+        ilComponentDefinitionProcessor ...$processor
+    ) {
+        $this->processor = $processor;
+    }
+
+    /**
+     * This methods is supposed to purge existing data in the registered
+     * processor.
+     */
+    public function purge() : void
+    {
+        foreach ($this->processor as $p) {
+            $p->purge();
+        }
+    }
+
+    /**
+     * This reads the component.xml of all components in the core and processes
+     * them with the provided processor.
+     */
+    public function readComponentDefinitions()
+    {
+        foreach ($this->getComponents() as list($type, $component, $path)) {
+            $file = $this->readFile($path);
+            foreach ($this->processor as $processor) {
+                $processor->beginComponent($component, $type);
+            }
+            $this->parseComponentXML($type, $component, $file);
+            foreach ($this->processor as $processor) {
+                $processor->endComponent($component, $type);
+            }
+        }
+    }
+
+    protected function readFile(string $path) : string
+    {
+        if (!file_exists($path)) {
+            throw new \InvalidArgumentException(
+                "Cannot find file $path."
+            );
+        }
+        return file_get_contents($path);
+    }
+
+    protected function parseComponentXML(string $type, string $component, string $xml)
+    {
+        $xml_parser = null;
+        try {
+            $xml_parser = xml_parser_create("UTF-8");
+            xml_parser_set_option($xml_parser, XML_OPTION_CASE_FOLDING, false);
+            xml_set_object($xml_parser, $this);
+            xml_set_element_handler($xml_parser, 'beginTag', 'endTag');
+            if (!xml_parse($xml_parser, $xml)) {
+                $code = xml_get_error_code($xml_parser);
+                $line = xml_get_current_line_number($xml_parser);
+                $col = xml_get_current_column_number($xml_parser);
+                $msg = xml_error_string($code);
+                throw new \InvalidArgumentException(
+                    "Error $code component xml of $type/$component, on line $line in column $col: $msg"
+                );
+            }
+        } finally {
+            if ($xml_parser) {
+                xml_parser_free($xml_parser);
+            }
+        }
+    }
+
+    public function beginTag($_, $name, $attributes)
```

Comment:
```suggestion
    public function beginTag(XMLParser $parser, string $name, array $attributes)
```

Reply:
The `XMLParser` hint does not seem to work. It is a `resource` we receive there at least in some case. Since I don't care about that param anyway, I did not put a typehint.